### PR TITLE
[REVIEW] Jit cache path improvements

### LINKED ISSUE #4973: [BUG] Raw boost error shown to user when parent directory of kernel cache path is read-only
**Describe the bug**
When running in an environment where the user's home directory is inaccessible or otherwise unavailable (e.g.: in a container environment on a cluster), cudf will throw a boost exception message to the user without any context, e.g.:
```
boost::filesystem::create_directories: Permission denied: "/home/.cudf"
```

**Steps/Code to reproduce bug**
Run `gtests/BINARY_TEST` or some other cudf sample code that will use JIT kernels in an environment where the home directory is unavailable or explicitly set `LIBCUDF_KERNEL_CACHE_PATH` to a missing and unwritable location (e.g.: `/foo`)

**Expected behavior**
I would expect one of the following:
* The error is ignored, proceeding as if the JIT cache is inaccessible (as it used to do when colliding with another user in `/tmp`), possibly emitting a single warning message when this occurs.
* The error message would indicate the error is related to the JIT cache and point the user to the `LIBCUDF_KERNEL_CACHE_PATH` environment variable as a potential workaround to the issue.

### LINKED ISSUE #4987: [FEA] use a cudf version directory even when user specifies JIT cache path
**Is your feature request related to a problem? Please describe.**
Currently when the user overrides the JIT cache path the kernels are deposited directly within the specified folder.  This can become problematic when crossing cudf versions, which is why cudf uses a version-specific directory when selecting the default JIT kernel cache path.  If the user's specified kernel cache path is not tied to the cudf version being used then there can be runtime issues with stale, incompatible kernels.

**Describe the solution you'd like**
cudf always creates a version-specific directory in the default path under the `.cudf` folder in the user's home directory.  I'd like cudf to consider `$HOME/.cudf` as the default kernel cache path and always create the cudf version directory (e.g.: `0.14`) underneath whatever kernel cache path was determined, whether that's the default path or user-specified.  Then users can specify a path that remains consistent across cudf versions and not have to worry about pairing the kernel paths and cudf versions to avoid weird runtime problems.

For example, if I specify `LIBCUDF_KERNEL_CACHE_PATH=/tmp/cudf-jlowe`, I'd like to see cudf deposit kernels under `/tmp/cudf-jlowe/0.14/` rather than directly in the `/tmp/cudf-jlowe` directory as it does today.

**Describe alternatives you've considered**
We can clearly document the correlation between the user's specified kernel cache path and the version of cudf being used.  That places a burden on users to update the setting across cudf versions which is avoided with the proposed approach.

### PR DESCRIPTION
**Pull request by devavret**
- Allows cache path access to fail. Falls back on no file cache. Closes #4973 
- Custom JIT cache path also contains cudf version directories to avoid inter version clashes. Closes #4987 
- Changed read file locks to shared to fix occasional hiccups in reading from NFS

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/jit/cache.cpp [hunk 1]
```diff
@@ -67,8 +68,14 @@ boost::filesystem::path getCacheDir()
   auto kernel_cache_path     = boost::filesystem::path(
     kernel_cache_path_env != nullptr ? kernel_cache_path_env : LIBCUDF_KERNEL_CACHE_PATH);
   if (not kernel_cache_path.empty()) {
```
Comment (vuule) [inline cpp/src/jit/cache.cpp:70]: When would `kernel_cache_path` be empty? The only scenario I see is someone setting the env `LIBCUDF_KERNEL_CACHE_PATH` to an empty string to disable caching. Any other cases?

File: cpp/src/jit/cache.cpp [hunk 2]
```diff
@@ -150,8 +157,13 @@ std::string cudfJitCache::cacheFile::read()
     return std::string();
   }
 
+  // Create args for file locking
+  flock fl{};
+  fl.l_type   = F_RDLCK;  // Shared lock for reading
+  fl.l_whence = SEEK_SET;
+
   // Lock the file descriptor. we the only ones now
```
Comment (vuule) [inline cpp/src/jit/cache.cpp:165]: This comment does not seem correct now, with the lock changes.

File: cpp/src/jit/cache.cpp (hunk 1 above)
Comment (devavret) [reply cpp/src/jit/cache.cpp:70]: The rare scenario where someone unsets the environment variable for HOME.

File: CHANGELOG.md [hunk 3]
```diff
@@ -195,6 +195,7 @@
 - PR #4965 Raise Error when there are duplicate columns sent to `cudf.concat`
 - PR #4984 Fix groupby nth aggregation negative n and exclude nulls
 - PR #4977 Fix compilation of cuDF benchmarks with build.sh
+- PR #5018 Miscellaneous Jit cache path improvements
```
[the remark this answers is not in the document]
Comment (devavret) [reply CHANGELOG.md:198]: The primary reason for this PR is the crash when the desired directory is not accessible. The other issue about having versioned directories in custom path is sort of a bug too

File: cpp/src/jit/cache.cpp (hunk 2 above)
Comment (devavret) [reply cpp/src/jit/cache.cpp:165]: You’re right. Thanks for the quick review

File: CHANGELOG.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (vuule) [reply CHANGELOG.md:198]: Then maybe it should be worded as such. Something like 'Fix crash when XYZ; other Jit cache fixes'

File: cpp/src/jit/cache.cpp (hunk 1 above)
Comment (vuule) [reply cpp/src/jit/cache.cpp:70]: Might be good to add a comment above this line.

**Pull request merged**